utils/binutils.ml

```diff
@@ -0,0 +1,577 @@
+    let buf = load_bytes d Int64.(add e_lfanew (of_int header_size)) size_of_optional_header in
+    let magic =
+      match get_uint16 d buf 0 with
+      | 0x10b -> PE32
+      | 0x20b -> PE32PLUS
+      | n -> raise (Error (Unsupported ("optional header magic", n)))
+    in
+    let image_base =
+      match magic with
+      | PE32 -> unsigned_of_int32 (get_uint32 d buf 28)
+      | PE32PLUS -> get_uint64 d buf 24
+    in
+    {magic; image_base}
+
+  type section =
+    {
+      name: string;
+      virtual_size: int;
+      virtual_address: int64;
+      size_of_raw_data: int;
+      pointer_to_raw_data: int64;
+    }
+
+  let section_header_size = 40
+
+  let read_sections d {e_lfanew; number_of_sections; size_of_optional_header; _} =
+    let buf =
+      load_bytes d
+        Int64.(add e_lfanew (of_int (header_size + size_of_optional_header)))
+        (number_of_sections * section_header_size)
+    in
+    let mk i =
+      let base = i * section_header_size in
+      let name = name_at ~max_len:8 buf (base + 0) in
+      let virtual_size = get_uint d buf (base + 8) in
+      let virtual_address = unsigned_of_int32 (get_uint32 d buf (base + 12)) in
+      let size_of_raw_data = get_uint d buf (base + 16) in
+      let pointer_to_raw_data = unsigned_of_int32 (get_uint32 d buf (base + 20)) in
+      {name; virtual_size; virtual_address; size_of_raw_data; pointer_to_raw_data}
+    in
+    Array.init number_of_sections mk
+
+  type symbol =
+    {
+      name: string;
+      address: int64;
+    }
+
+  let load_section_body d {size_of_raw_data; pointer_to_raw_data; _} =
+    load_bytes d pointer_to_raw_data size_of_raw_data
+
+  let find_section sections sectname =
+    array_find (function ({name; _} : section) -> name = sectname) sections
+
+  let read_symbols d {image_base; _} sections =
+    match find_section sections ".exptbl" with
+    | None -> [| |]
+    | Some ({virtual_address; _} as exptbl) ->
+        let buf = load_section_body d exptbl in
+        let numexports = Int64.to_int (get_word d buf 0) in
+        let word_size = word_size d in
+        let mk i =
+          let address = get_word d buf (word_size * (2 * i + 1)) in
+          let nameoff = get_word d buf (word_size * (2 * i + 2)) in
+          let name =
+            let off = Int64.(sub nameoff (add virtual_address image_base)) in
+            name_at buf (Int64.to_int off)
+          in
+          {name; address}
+        in
+        Array.init numexports mk
+
+  let symbol_offset {image_base; _} sections symbols =
+    match find_section sections ".data" with
+    | None -> Fun.const None
+    | Some {virtual_address; pointer_to_raw_data; _} ->
+        fun symname ->
+          begin match array_find (function {name; _} -> name = symname) symbols with
+          | None -> None
+          | Some {address; _} ->
+              Some Int64.(add pointer_to_raw_data (sub address (add virtual_address image_base)))
```

Comment:
Same as above for the address mapping. Unclear to me if that's correct because I don't know about flexdll's semantics.

Reply:
See above.